### ISSUE TITLE
Long TCP sessions lose their UFT entries

### ISSUE THREAD

**Issue by FelixMcFelix** [oxidecomputer/opte#426]
@citrus-it found that [ZFS transfers from a DigitalOcean VM to Colo](https://gist.github.com/citrus-it/fef4602f4c0494efd188a990665fd4e7) are failing with some odd characteristics -- transfers failed around 3.98GB completion, while outbound traffic on the connection (e.g., ACKs) successfully transits Colo->DO.

```
╔═════════════╗                   ╔═════════════╗
║             ╠────────ACKs───────▶             ║
║    Colo     ║                   ║DigitalOcean ║
║     Vm      ║             TCP   ║     VM      ║
║             ◀ ─ ─ ─ ─ X◀──Segs──╣             ║
╚═════════════╝                   ╚═════════════╝
```

We spent some time digging into this, broadly finding:
* NAT mappings are a-okay (`swadm nat get -e <ip> -p 22`), although they were invisible via `swadm nat list`.
* Issue persists after sled migration.
* Issue does *not* present itself when using iperf3 to send traffic from DO->Colo, but is dead reproducible with a zfs copy.
* Failure is not explicitly tied to size, but seems to happen at or around 30s boundaries (3:00, 3:30, 4:00, ...) [i.e., up to around 9GB].

Focussing on OPTE, we found that the UFT-In entry for the above flow is suddenly *expired* according to DTrace, in spite of its hit count actively increasing. After this point, no segments reach the Colo VM in spite of retransmits (no replacement UFT entry built). Expiry in particular is **very weird**: our understanding is that the packet <1s before expiry should have updated the `FlowEntry`'s `last_hit`. Our gut feelings seemed to indict either the TCP state tracking and/or `Moment::delta_as_millis()` -- while `ddi::gethrtime()` is monotonic, it was suggested that there might be per-core variation.

**Comment by citrus-it** [comment]
Here's an example of an active outbound UFT entry going away. This corresponds with the `flow-expired` USDT probe firing. One can see that the hit count is increasing right up to the point that the flow disappears, so the open question is around why it was deemed to have expired.

```
Tue Dec 12 15:14:41 UTC 2023

UFT Inbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    164.90.148.125   22     45.154.216.124   38174  7365157  hdr: decap,NAT,, body:

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4076079  hdr: NAT,encap, body:

Tue Dec 12 15:14:42 UTC 2023

UFT Inbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    164.90.148.125   22     45.154.216.124   38174  7403265  hdr: decap,NAT,, body:

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4095611  hdr: NAT,encap, body:

Tue Dec 12 15:14:43 UTC 2023

UFT Inbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    164.90.148.125   22     45.154.216.124   38174  7441371  hdr: decap,NAT,, body:

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4115162  hdr: NAT,encap, body:

Tue Dec 12 15:14:44 UTC 2023
UFT Inbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    164.90.148.125   22     45.154.216.124   38174  7479534  hdr: decap,NAT,, body:

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4134720  hdr: NAT,encap, body:

Tue Dec 12 15:14:45 UTC 2023

UFT Inbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    164.90.148.125   22     45.154.216.124   38174  7517632  hdr: decap,NAT,, body:

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4154239  hdr: NAT,encap, body:

Tue Dec 12 15:14:46 UTC 2023

UFT Inbound: 2/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS

UFT Outbound: 3/65535
----------------------------------------------------------------------
PROTO  SRC IP           SPORT  DST IP           DPORT  HITS     XFORMS
TCP    172.30.0.5       38174  164.90.148.125   22     4161882  hdr: NAT,encap, body:
```

**Comment by rcgoodfellow** [comment]
> Issue does not present itself when using iperf3, but is dead reproducible with a zfs copy

What direction was the iperf session in e.g, who was the client and who was the server?

**Comment by citrus-it** [comment]
> What direction was the iperf session in e.g, who was the client and who was the server?

Inbound to the colo VM from a client out in DO.

I have now successfully reproduced this on dogfood so will continue investigating there.

**Comment by citrus-it** [comment]
I can trigger this almost instantly on dogfood if I use dtrace to blow up the suspected race window.

```
BRM42220006 # dtrace -wn 'fbt:xde:*expire_flows*:entry{chill(25000000)}'
dtrace: description 'fbt:xde:*expire_flows*:entry' matched 5 probes
dtrace: allowing destructive actions

BRM42220006 # dtrace -L ./lib -I . -Cqs ./opte-flow-expire.d
PORT                     FT NAME            FLOW LAST_HIT NOW
opte1                    uft_in             TCP,172.20.3.69:22,172.20.26.118:38896 0 0
opte1                    uft_out            TCP,172.30.0.10:38896,172.20.3.69:22 0 0
```

**Comment by FelixMcFelix** [comment]
Fab! I think that seems pretty damning. For documentation, the suspected race window in question is down to computing `Moment::now()` before acquiring the port lock for expiry:

1. `xde::expire_periodic` computes `Moment::now()` at time $t_0$ before calling into `port.expire_flows()`.
2. A packet in either direction acquires the port lock for processing (`self.data.lock()`).
3. `Port::expire_flows` attempts to acquire the port lock and waits.
4. On inbound packet processing, UFT's `last_hit` is set to $t_1 :=$ `Moment::now()` $> t_0$.
5. Packet processing completes and port lock dropped. Periodic task resumes.
6. Time since UFT was last updated is computed as `now - last_hit` $=t_0-t_1$. As $t_0 < t_1$, we get a `u64` wrap and the flow is expired.

Saturating to zero should prevent the wrap (#427), but either way we should really also fix the timestamp acquisition.

**Comment by FelixMcFelix** [comment]
Issue appears to be remedied on colo; full data transfer was completed without a hitch. While #427 is *a* fix for this problem, I've opened #434 which addresses the actual timestamp generation.
